dedalus/core/problems.py

```diff
@@ -329,201 +170,165 @@ def _build_matrix_expressions(self, eqn):
             F['c'] = 0
+        # Save expressions and metadata
         eqn['L'] = L
         eqn['F'] = F
+        eqn['domain'] = domain
         eqn['matrix_dependence'] = L.matrix_dependence(*vars)
         eqn['matrix_coupling'] = L.matrix_coupling(*vars)
-        #eqn['separability'] = eqn['L'].separability(*vars)
         # Debug logging
-        logger.debug('  {} linear form: {}'.format('L', eqn['L']))
+        logger.debug(f"  L: {L}")
+        logger.debug(f"  F: {F}")
 
 
 class NonlinearBoundaryValueProblem(ProblemBase):
     """
-    Class for nonlinear boundary value problems.
+    Nonlinear boundary value problems.
 
     Parameters
     ----------
-    domain : domain object
-        Problem domain
+    variables : list of Field objects
+        Problem variables to solve for.
+    namespace : dict-like, optional
+        Dictionary for namespace additions to use when parsing strings as equations
+        (default: None). It is recommended to pass "locals()" from the user script.
 
     Notes
     -----
-    This class supports nonlinear boundary value problems.  The LHS and RHS
-    terms are recombined to form a root-finding problem:
-
-        F(X) = G(X)  -->  H(X) = F(X) - G(X) = 0
+    This class supports nonlinear boundary value problems of the form:
+        F(X) = G(X)
+    which are recombined to form the root-finding problem:
+        H(X) = F(X) - G(X) = 0
 
     The problem is reduced into a linear BVP for an update to the solution
-    using the Newton-Kantorovich method and symbolically-computed Frechet
-    derivatives of the equation.
-
-        H(X0 + X1) = 0
-        H(X0) + dH(X0).X1 = 0
-        dH(X0).X1 = -H(X0)
-        Solve for X1
-
+    using the Newton-Kantorovich method and the symbolically-computed Frechet
+    differential of the equation:
+        H(X[n+1]) = 0
+        H(X[n] + dX) = 0
+        H(X[n]) + dH(X[n]).dX = 0
+        dH(X[n]).dX = - H(X[n])
+
+    Iteration procedure:
+        - Form dH(X[n])
+        - Evaluate H(X[n])
+        - Solve dX = - dH(X[n]) \ H(X[n])
+        - Update X[n+1] = X[n] + dX
     """
 
     solver_class = solvers.NonlinearBoundaryValueSolver
 
     def __init__(self, *args, **kw):
         super().__init__(*args, **kw)
+        # Build perturbation variables
         self.perturbations = [var.copy() for var in self.variables]
         for pert, var in zip(self.perturbations, self.variables):
             pert['c'] = 0
             pert.name = 'δ'+var.name
         self.LHS_variables = self.perturbations
 
-    @CachedAttribute
-    def namespace_additions(self):
-        """Build namespace for problem parsing."""
-        # Variable perturbations
-        self.perturbations = [Field(bases=var.bases, name='δ'+var.name, dtype=var.dtype) for var in self.variables]
-        return {pert.name: pert for pert in self.perturbations}
+    def _check_equation_conditions(self, eqn):
+        """Check equation conditions."""
+        # No conditions
```

Comment:
Do we need to check linearity of LHS?

Reply:
Maybe also domain containment?

Reply:
We don't need them since the current NLBVP setup takes `LHS = RHS` and converts it to a straight nonlinear root-finding using both sides: `H(X) = LHS - RHS = 0`, so there's no restrictions on LHS vs RHS.